geode-core/src/main/java/org/apache/geode/internal/cache/xmlcache/CacheCreation.java

```diff
@@ -521,12 +521,12 @@ void create(InternalCache cache)
 
     cache.initializePdxRegistry();
 
-    for (DiskStore diskStore : diskStores.values()) {
+    diskStores.values().parallelStream().forEach(diskStore -> {
       DiskStoreAttributesCreation creation = (DiskStoreAttributesCreation) diskStore;
       if (creation != pdxRegDSC) {
         createDiskStore(creation, cache);
       }
-    }
+    });
```

Comment:
How is the exception handled?

Reply:
Need to add unit/integration tests for both cases, with multiple and single thread trying to create disk-stores.

Reply:
I am not sure which exception you are talking about. I don't see `parallelStream()` and `forEach()` introduce new exception, based on the Javadoc.

For test coverage, there are quite a few existing tests that already cover the single thread or multiple disk stores cases. We don't have to add new tests for that.

For multi-thread test, it is a very tricky one. The purpose of this pull request is to introduce parallel disk recovery and improve the performance. We can assert that the stream should be parallel using `isParallel()`. Or we can scan the logs and make sure there are multiple thread IDs that recover the disk stores. But it doesn't serve the purpose of testing performance improvement. It doesn't guarantee that disk recovery is parallel, even if we can assert that the stream is parallel or there are multiple threads that recover the disk store. For example, if we keep the synchronized block in `DiskStoreFactoryImple` line 147, the disk stores recovery will be sequential, even though there are multiple threads recovering the disk stores.

Reply:
Based on the DiskStoreFactory.create() it can throw IllegalArgumentException; probably IOException based on the disk read failures. In case there are multiple disk-store failures, we will leave the parallel stream to throw the exception or we will capture all the exception and throw a single exception with all the disk=store names in it...

I was thinking from unit test level; if there are already tests covering the new code. No need of new tests.

Reply:
The system property is added. Tests are added as well.

Reply:
Can we move this code block to helper method...The create() is becoming too long...

Reply:
How about test where one/two of the recovery thread throws IllegalArgumentException...

Reply:
Refactored the code block to helper method. Done.

Reply:
I am not sure which method throws `IllegalArgumentException`. To the best of my knowledge, the new code I introduced does not throw `IllegalArgumentException`. If existing code throws `IllegalArgumentException`, it is beyond the scope of this JIRA.